src/protocol/attribute.rs

```diff
@@ -0,0 +1,224 @@
+use super::{AttributionModel, RecordId};
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::mesh::{Gateway, Mesh},
+    protocol::context::ProtocolContext,
+    secret_sharing::Replicated,
+};
+use futures::future::try_join_all;
+
+/// Obilivious attribution protocol
+///
+/// Obliviously attribute trigger events to corresponding source events for different breakdowns in a MPC.
+/// There are several common attribution models such as Last Touch, Even Credit and Exponential Decay.
+/// https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#oblivious-last-touch-attribution
+#[allow(dead_code)]
+pub struct Attribute<'a, G> {
+    ctx: &'a ProtocolContext<'a, G, AttributionModel>,
+    step: AttributionModel,
+}
+
+impl<'a, G> Attribute<'a, G> {
+    const _IS_TRIGGER_BIT: usize = 2;
+    const _HELPER_BIT: usize = 3;
+    const _VALUE: usize = 5;
+
+    #[allow(dead_code)]
+    pub fn new(ctx: &'a ProtocolContext<'a, G, AttributionModel>, step: AttributionModel) -> Self {
+        Self { ctx, step }
+    }
+
+    /// Executes the protocol and produces an attribution result for the model specified by `step` in `Attribute::new`.
+    /// Attribution result will be used to calculate the aggregation in the next step.
+    #[allow(dead_code)]
+    pub async fn execute<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        match self.step {
+            AttributionModel::LastTouch => self.last_touch_attribution(input).await,
+        }
+    }
+
+    async fn last_touch_attribution<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        // TOOD: input length range check. Can we assume `assert!(num_rows < u32::MAX)`?
+        let num_rows = input[0].len();
+
+        // 1. Create Credit and Stop bit vectors
+        let mut credits: Vec<Replicated<F>> = Vec::with_capacity(num_rows);
+        input[Self::_VALUE].iter().for_each(|x| credits.push(*x));
+
+        let one = self.ctx.gateway.get_channel(self.step).share_of_one::<F>();
+        let mut stop_bits: Vec<Replicated<F>> = vec![one; num_rows];
+
+        // 2. Accumulate
+        //
+        // Last touch accumulation algorithm:
+        //     b = stop_bit * successor.helper_bit * successor.trigger_bit;
+        //     credit = credit + b * successor.credit;
+        //     stop_bit = b * successor.stop_bit;
+        let mut step_size = 1;
+        while step_size < num_rows {
+            let end = num_rows - step_size;
+
+            // first, calculate [successor.helper_bit * successor.trigger_bit]
+            let mut b = self
+                .multiply_all(
+                    &input[Self::_HELPER_BIT][step_size..],
```

Comment:
I find this ordering of the arrays surprising.  Why do we have an array of arrays where the outer array is indexed by the field type?  Why are we using array indices to access shares/fields anyway?

I had expected that we would be operating on a sequence of structs, where each field in the struct is a replicated share.  This looks like a sequence of sequences, where the outer sequence has some sort of implicit structure.  That leads to some odd potential failure modes that need to be guarded against, like where the Vec for the helper bit is a different length to the Vec for the stop bit.

Reply:
I was just discussing this with Richa and others. I agree that we expect to receive a sequence of structs. When it comes to the actual protocol executions, we realized we were always processing a sequence of replicated shares for a certain field (i.e. match key). For this PR, I chose this input format for the sake of simplicity, assuming the input had been transposed. Should all top-level protocol steps expect that the input format would be a sequence of structs?

Reply:
I'm confused by your reply taiki.

I think Martin is suggesting something like:

struct AttributionInputs {
    helper_bit: Replicated<F>,
    is_trigger_bit: Replicated<F>,
    value: Replicated<F>,
    breakdown_key: Replicated<F>,
}

And the "execute" function would accept a Vec of these things.

Reply:
The code was bit more readable to write

```
.multiply_all(
    &self.input[Self::_HELPER_BIT][step_size..],
    &self.input[Self::_IS_TRIGGER_BIT][step_size..],
)
```

than

```
impl Index for AttributionInputs { ... }

.multiply_all(
    &self.input[step_size..],
    SELF::_HELPER_BIT,
    SELF::_IS_TRIGGER_BIT,
)
```

that's all. I'm okay with the latter.

As for the sort protocol, I believe it creates a vector for each sort key. @richajaindce, your thoughts?

Reply:
It looks like we always operate on same type. This is true for running attribution and also if we want to execute a shuffle data attached to matchkey before sort (to hide information).  Would keeping all of data in memory be an issue? Using Taiki's approach, not everything needs to be in memory vs event approach needs all data to be in memory.

Reply:
I am going to fix this part to take a slice of structs, but just wanted to clarify what's going on in the code.
So, the [attribution algorithm](https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#oblivious-last-touch-attribution) would look like this:

```
// input: Vec<GenericReport>
let mut credit: Vec<Replicated<F>> = input.map(|x| x.value).collect();
let mut stop_bit: Vec<Replicated<F>> = [F::ONE; input.len()];

while step_size < input.len()
    let mut new_credit = credit.clone();
    let mut new_stop_bit = stop_bit.clone();

    for i in 0..input.len() {
        if i+step_size > input.len() break;

        let current = i;
        let successor = i+step_size;

        // in reality these will be lengthy async multiply calls
        let flag = stop_bit[i] * input[successor].helper_bit * input[successor].trigger_bit;
        new_credit[current] = credit[current] + flag * credit[successor];
        new_stop_bit[current] = flag * stop_bit[successor];
    }

    // `new_credit` is (input.len() - step_size) long, but `credit` must be input.len() length
    let end = input.len() - step_size;
    credit = new_credit[0..end].append(credit[end..]);
    // same for stop_bit
    stop_bit = new_stop_bit[0..end].append(stop_bit[end..]);
    
    step_size *= 2;
}
...
// we drop unnecessary fields. only need to return aggregated credits
Ok([{breakdown_key, credit}; unique(breakdown_key).len()])
```

Note that a `for` loop iteration doesn't depend on the previous iterations. Since they are independent we can eliminate the `for` loop and execute multiplications on slices of shares concurrently, which is what `multiply_all` does. It also eliminates awkward boundary checks, reassign of variables, padding vectors, etc.